pkg/controller/runtime/runtime.go

```diff
@@ -240,32 +246,133 @@ func (runtime *Runtime) watch(resourceNamespace resource.Namespace, resourceType
+	// to the second goroutine.
+	//
+	// There is a trick being used which sends a single map back and forth between the two goroutines.
+	// There is no locking required, as the map is owned by a single goroutine at a single moment of time.
+	// Additional channel 'empty' is used to block the second goroutine when there are no events to process.
+	ch := make(chan dedup, 1)
+	empty := make(chan dedup, 1)
+	empty <- dedup{}
+
+	go runtime.deduplicateWatchEvents(ch, empty)
+	go runtime.deliverDeduplicatedEvents(ch, empty)
+}
+
+// deduplicateWatchEvents deduplicates events from the watch channel into the map sent to the channel ch.
+func (runtime *Runtime) deduplicateWatchEvents(ch chan dedup, empty <-chan dedup) {
 	for {
-		var e state.Event
+		var events []state.Event
+
+		// wait for an event
+		select {
+		case <-runtime.runCtx.Done():
+			return
+		case events = <-runtime.watchCh:
+		}
+
+		// acquire a map
+		var m dedup
 
 		select {
+		case m = <-empty:
+		case m = <-ch:
 		case <-runtime.runCtx.Done():
 			return
-		case e = <-runtime.watchCh:
 		}
 
-		if e.Type == state.Errored {
-			// watch failed, we need to abort
-			runtime.watchErrors <- e.Error
+		processEvents := func() bool {
+			for _, e := range events {
+				if e.Type == state.Errored {
+					// watch failed, we need to abort
+					runtime.watchErrors <- e.Error
+
+					return false
+				}
+
+				m[reduceMetadata(e.Resource.Metadata())] = struct{}{}
+			}
 
+			return true
+		}
+
+		if !processEvents() {
 			return
 		}
 
-		md := e.Resource.Metadata()
+		// drain the watchCh by consuming all immediately available events
+	drainer:
+		for {
+			select {
+			case events = <-runtime.watchCh:
+				if !processEvents() {
+					return
+				}
+			case <-runtime.runCtx.Done():
+				return
+			default:
+				break drainer
+			}
+		}
+
+		// send the map to the second goroutine for processing
+		if !channel.SendWithContext(runtime.runCtx, ch, m) {
+			return
+		}
+	}
+}
+
+// deliverDeduplicatedEvents delivers events from the deduplicated channel to the controllers.
+func (runtime *Runtime) deliverDeduplicatedEvents(ch chan dedup, empty chan<- dedup) {
+	for {
+		// wait for a map
+		var m dedup
+		select {
+		case m = <-ch:
+		case <-runtime.runCtx.Done():
+			return
+		}
+
+		// consume any first key of the map
+		var k reducedMetadata
+
+		for k = range m {
+			break
+		}
+
+		delete(m, k)
+
+		// send the map back to the first goroutine
+		if len(m) > 0 {
+			if !channel.SendWithContext(runtime.runCtx, ch, m) {
+				return
+			}
+		} else {
+			if !channel.SendWithContext(runtime.runCtx, empty, m) {
+				return
+			}
+		}
 
+		// notify controllers
 		controllers, err := runtime.depDB.GetDependentControllers(controller.Input{
-			Namespace: md.Namespace(),
-			Type:      md.Type(),
-			ID:        pointer.To(md.ID()),
+			Namespace: k.namespace,
+			Type:      k.typ,
+			ID:        pointer.To(k.id),
 		})
 		if err != nil {
 			// TODO: no way to handle it here
```

Comment:
I think we can at least log it using `Runtime.logger`. This comment predates that field, so maybe we can change this too?

Reply:
fixed